Improve error handling in --json mode

mikix opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------

The main problem this solves are:
1. The restore command does not provide JSON versions of its errors.
2. The backup command's error output in --json mode does not actually show the error. Because the json package [can't marshal an error](https://github.com/golang/go/issues/5161).

This addresses both of those, as well as a few fixes for issues I found along the way:
- The backup command was printing scanner errors to stderr, not stdout.
- Added a ui.Terminal interface, to make it easier to test printing code.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
Closes #4944
Closes #4945

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

mikix committed:
backup: print scanner errors to stderr, not stdout

mikix committed:
backup: show actual error strings in --json mode

Previously, an error JSON fragment would look like:
{"message_type": "error", "error": {}}

This is because encoding/json cannot marshal an error interface.
Instead, we now call .Error() to get the string value.

mikix committed:
restore: print JSON versions of errors in --json mode

Previously, they were printed as freeform text.

This also adds a ui.Terminal interface to make writing
tests easier and also adds a few tests.

mikix commented on internal/ui/restore/json.go:
```diff
@@ -99,6 +113,12 @@ type statusUpdate struct {
 	BytesSkipped   uint64  `json:"bytes_skipped,omitempty"`
 }
 
+type errorUpdate struct {
+	MessageType string `json:"message_type"` // "error"
+	Error       string `json:"error"`
+	Item        string `json:"item"`
+}
```
Should this include a `During` field, like the backup errors do? Restore doesn't have different phases like the backup command does, but I wondered if consistency of JSON blobs across commands was valuable enough to include a `"during": "restore"` value (or similar).

mikix commented on internal/restic/node_test.go:
```diff
@@ -382,3 +383,14 @@ func TestSymlinkSerializationFormat(t *testing.T) {
 	}
 }
+
+func TestNodeRestoreMetadataError(t *testing.T) {
```
I would have added more interesting tests here, but mocking sys calls is a little awkward in Go, I found.

MichaelEischer commented on internal/ui/backup/text_test.go:
```diff
@@ -0,0 +1,47 @@
+	"github.com/restic/restic/internal/test"
+)
+
+type mockTerm struct {
```
That `mockTerm` looks identical to the one used in `ui/restore`. Instead of duplicating them, please move it to `internal/ui/mock.go` (we could also create a subfolder, but that feels like overkill).

MichaelEischer commented on internal/ui/termstatus/status.go:
```diff
@@ -15,6 +15,13 @@ import (
 	"golang.org/x/text/width"
 )
 
+type TerminalWriter interface {
```
Let's move this struct to `internal/ui/terminal.go` (or similar). This will finally allow us to stop importing `ui/termstatus` in `internal/ui/message.go`.

MichaelEischer reviewed (commented):
Thanks for the great PR! I have a few smaller comments, see below.

I'm a bit undecided how to handle the broken error message in the backup JSON output. It might be a good idea to maintain backwards compatibility with implementations that for example directly use the Go struct from the source code. Decoding the struct in Go would AFAICT result in a type error, when trying to decode `{"error":"some message"}` instead of `{"error":{}}`. So it might be better to use `{"error":{"message":"some value"}}` instead. That leads to a bit of not entirely necessary nesting, but would be fully backwards compatible and also provide room for adding an error code or similar. I'll have to think a bit about that.

[Edit]Parsing an old error message would fail: https://go.dev/play/p/qnXMavntl02 [/Edit]

mikix commented on internal/ui/backup/text.go:
```diff
@@ -73,7 +73,7 @@ func (b *TextProgress) Update(total, processed Counter, errors uint, currentFile
 // error in verbose mode and returns nil.
 func (b *TextProgress) ScannerError(_ string, err error) error {
-	b.V("scan: %v\n", err)
+	b.E("scan: %v\n", err)
```
Good point - I can verbosity-guard the `E()` call here.

mikix commented on internal/ui/backup/json.go:
```diff
@@ -208,7 +208,7 @@ type statusUpdate struct {
 type errorUpdate struct {
 	MessageType string `json:"message_type"` // "error"
-	Error       error  `json:"error"`
+	Error       string `json:"error"`
```
Based on your Go playground test, sounds like there's not much value in trying to be backwards compatible here?

i.e. I might as well leave this part of the PR as-is and just re-use the "error" field as a string instead of an an object.

MichaelEischer replied:
I think it's the other way around. My go playground example shows that if we change `error` to `string`, then it will be a breaking change. Whereas the inline struct variant remains backwards compatible. Having a nested struct isn't entirely ideal, but could be worse. So let's follow the approach of turning the error field into a struct.

For consistency that should also apply to the `restore` command.

mikix replied:
Got it, will do :+1:

mikix replied:
Done.

mikix commented on internal/restic/node.go:
```diff
@@ -239,33 +246,26 @@ func (node Node) restoreMetadata(path string, warn func(msg string)) error {
 	if err := node.RestoreTimestamps(path); err != nil {
 		debug.Log("error restoring timestamps for dir %v: %v", path, err)
-		if firsterr != nil {
+		if firsterr == nil {
```
Filed https://github.com/restic/restic/pull/4958 - will update this PR in a bit with the commit removed and the rest of your comments addressed.

mikix commented:
I've updated this PR to remove the metadata-error change (now in a separate PR) and address all your comments, I believe. Please re-review! Thanks

MichaelEischer commented:
It will probably take a few days until I can take a look. I'll first have to sort out the regressions from restic 0.17.0.

mikix commented:
No rush! I know I’m throwing a lot of PRs at ya, but there’s no time pressure from my end. Regressions come first! 😄

MichaelEischer reviewed (commented):
The changes look good so far. The only thing left to sort out are the linter errors and the replacement type for the `error` field.

mikix committed:
json: switch backup and restore errors from string to struct types

This keeps backwards compatibility with the previous empty structs.
And maybe we'd want to put other fields into the inner struct later,
rather than the outer message.

mikix commented:
OK should be ready again, and after noticing your comment about `doc/075_scripting.rst` in another PR, I updated that file here too.

MichaelEischer reviewed (approved):
LGTM. Thanks!
